ffcx/codegeneration/expressions.py

```diff
@@ -227,34 +227,67 @@ def generate_block_parts(self, blockmap, blockdata):
 
-        # Prepend dimensions of dofmap block with free index
-        # for quadrature points and expression components
-        B_indices = tuple([iq] + list(arg_indices))
-
-        # Fetch code to access modified arguments
-        # An access to FE table data
-        arg_factors = self.get_arg_factors(blockdata, block_rank, B_indices)
-
-        A_indices = []
-        for i in range(len(blockmap)):
-            offset = blockmap[i][0]
-            A_indices.append(arg_indices[i] + offset)
-        A_indices = tuple([iq] + A_indices)
-
-        # Multiply collected factors
-        # For each component of the factor expression
-        # add result inside quadloop
-        body = []
-
-        for fi_ci in blockdata.factor_indices_comp_indices:
-            f = self.get_var(F.nodes[fi_ci[0]]["expression"])
-            Brhs = L.float_product([f] + arg_factors)
-            body.append(L.AssignAdd(A[(fi_ci[1],) + A_indices], Brhs))
-
-        for i in reversed(range(block_rank)):
-            body = L.ForRange(
-                B_indices[i + 1], 0, blockdims[i], body=body)
-        quadparts += [body]
+        # Check if DOFs in dofrange are equally spaced
+        unzip = False
+        for i, bm in enumerate(blockmap):
+            for a, b in zip(bm[1:-1], bm[2:]):
+                if b - a != bm[1] - bm[0]:
+                    unzip = True
+                    break
+            else:
+                continue
+            break
+
+        if unzip:
+            # If DOFs in dofrange are not equally spaced
+            from itertools import product
+            for A_indices, B_indices in zip(product(*blockmap),
+                                            product(*[range(len(b)) for b in blockmap])):
+                B_indices = tuple([iq] + list(B_indices))
+                A_indices = tuple([iq] + A_indices)
+                for fi_ci in blockdata.factor_indices_comp_indices:
+                    f = self.get_var(F.nodes[fi_ci[0]]["expression"])
+                    arg_factors = self.get_arg_factors(blockdata, block_rank, B_indices)
+                    Brhs = L.float_product([f] + arg_factors)
+                    quadparts.append(L.AssignAdd(A[(fi_ci[1],) + A_indices], Brhs))
+        else:
+
+            # Prepend dimensions of dofmap block with free index
+            # for quadrature points and expression components
+            B_indices = tuple([iq] + list(arg_indices))
+
+            # Fetch code to access modified arguments
+            # An access to FE table data
+            arg_factors = self.get_arg_factors(blockdata, block_rank, B_indices)
+
+            # TODO: handle non-contiguous dof ranges
+
+            A_indices = []
+            for bm, index in zip(blockmap, arg_indices):
+                # TODO: switch order here? (optionally)
+                # TODO: save block size in ir rather than reverse engineering here
+                offset = bm[0]
+                if len(bm) == 1:
+                    ebs = 1
```

Comment:
What is `ebs`? Why are you reverse engineering block size here, it is stored in IR.

Reply:
block size isn't stored in the integral/expression ir, so I'm not sure you can access it here. Also, this allows for elements that have gaps in their dof lists for reasons other than blocksize (which there is probably no current use case, but doing it this way allows for this in future)